### PR TITLE
Added additional event sources for generate-event

### PR DESCRIPTION
**Pull request by stshan**
Added some additional event sources and event samples for generate-event

### CODE REVIEW

File: templates/AWS/ApiGatewayAuthorizer.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (sanathkr) [inline templates/AWS/ApiGatewayAuthorizer.json]: `template` usually refers to SAM templates. Can you instead call the folder `sample-events` or `sample-inputs` or `input-events` or something that sounds better?

Comment (stshan) [reply templates/AWS/ApiGatewayAuthorizer.json]: done

**Comment by stshan** [comment]
After pushing I realized that the explanation probably could be a bit more detailed. I'll make another change going more in-depth

**Comment by stshan** [comment]
As more event sources are added, it's pretty obvious that we can't keep doing what we're doing now (i.e. just keep appending more json to the generate-event file), so this change enables automatic generate-event command generation as you add more json files to the sample-events folder.

caveats: Since go doesn't let you package asset files into the binary by default, the json data needs to be converted into a .go file. I used https://github.com/jteeuwen/go-bindata to do this. However, this does mean that whenever any changes are made to the sample-events folder, the generated binary go file will need to be updated by running 'go generate'.

File: generate-event.go [hunk 2]
```diff
@@ -1,367 +1,158 @@
...
+		categories = append(categories, command)
 	}
 
-	switch eventType {
-	case "S3":
+	return
+}
+
+func generateCategorySubCommands(subDir string) (subcommands []cli.Command){
+	path := "sample-events/" + subDir
```
Comment (sanathkr) [inline generate-event.go]: this slash won't work in Windows

File: generate-event.go [hunk 3]
```diff
@@ -1,367 +1,158 @@
...
+	for _, file := range files {
+		name := strings.Split(file, ".")[0]
 
-	case "SNS":
+		command := cli.Command{
+			Name:  nameToCommandMap[name],
+			Usage: "Generates a sample " + nameToDescriptionMap[name] + " event",
+			Flags: generateCommandFlags(name),
+			Action: func(c *cli.Context) {
+				b, err := Asset(path + "/" + name + ".json") // just pass the file name
```
Comment (sanathkr) [inline generate-event.go]: slash - windows support. Python provides a os.path.join method that can construct a path. Go should have a similar one that does the joining with right slashes for right platform

File: sample-events/event-mapping.json [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    },
+    {
+      "name": "ApiGatewayAuthorizer",
+      "command": "api-authorizer",
+      "description": "API Gateway Authorizer",
+      "flags": []
+    },
+    {
+      "name": "ApiGatewayAwsProxy",
+      "command": "api-aws-proxy",
```
Comment (sanathkr) [inline sample-events/event-mapping.json]: this should be called `api` for backwards compatibility.

File: sample-events/event-mapping.json [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        {
+          "name": "key, k",
+          "usage": "The S3 key the event should reference",
+          "value": "test/key"
+        }
+      ]
+    },
+    {
+      "name": "S3Put",
+      "command": "put",
```
Comment (sanathkr) [inline sample-events/event-mapping.json]: Is this actually a "subcommand" then?

File: sample-events/event-mapping.json [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        {
+          "name": "key, k",
+          "usage": "The S3 key the event should reference",
+          "value": "test/key"
+        }
+      ]
+    },
+    {
+      "name": "ScheduledEvent",
+      "command": "scheduled-event",
```
Comment (sanathkr) [inline sample-events/event-mapping.json]: This was called `schedule`. Is it possible to change the name of all existing events to match the original command for back compat?

File: generate-event.go [hunk 7]
```diff
@@ -1,367 +1,158 @@
...
+
+	dirs, err := AssetDir("sample-events")
 	if err != nil {
-		fmt.Printf("Failed to load sample %s event: %s", eventType, err)
-		os.Exit(1)
+		log.Fatal(err)
+	}
+
+	for _, dir := range dirs {
+		if (strings.Contains(dir, ".")) {
```
Comment (sanathkr) [inline generate-event.go]: hmm.. you should use Go's neat APIs to do this check for you - https://stackoverflow.com/questions/8824571/golang-determining-whether-file-points-to-file-or-directory

File: generate-event.go [hunk 8]
```diff
@@ -1,367 +1,158 @@
...
+	Value   string `json:"value"`
+}
 
-	t, err := template.New("event").Parse(sample)
+// Generates top-level commands for generate-event for all
+// categories under sample-events
+func generateEventCategories() (categories []cli.Command) {
+	mapFileToCommands()
+
+	dirs, err := AssetDir("sample-events")
```
Comment (sanathkr) [inline generate-event.go]: can you make `sample-events` of this a global constant?

File: generate-event.go [hunk 9]
```diff
@@ -1,367 +1,158 @@
...
 	}
 
-	switch eventType {
-	case "S3":
+	return
+}
+
+func generateCategorySubCommands(subDir string) (subcommands []cli.Command){
+	path := "sample-events/" + subDir
+	files, err := AssetDir(path)
```
Comment (sanathkr) [inline generate-event.go]: `AssetDir` is a method from go-bindata right? You don't seem to include it here why?

Comment (stshan) [reply generate-event.go]: The function is from the generated binfile, and not from go-bindata, since go-bindata is just a tool and not a library

Comment (sanathkr) [reply generate-event.go]: Fair enough. So `generate-event-binary.go` is the file that provides these methods. It means ppl should `go generate` at least once before running `go build`, correct? In that case, is this section accurate? https://github.com/awslabs/aws-sam-local#build-from-source If not, can you update it?

Comment (stshan) [reply generate-event.go]: Yeah, I can update that. I do have a section under sample-events specifically for this, but I think it's probably better to mention it to the main README as well

https://github.com/stshan/aws-sam-local/tree/develop/sample-events

Comment (sanathkr) [reply generate-event.go]: sure thanks

**Comment by sanathkr** [comment]
With v0.3.0, this PR is not relevant anymore

**Pull request closed without merging**